README.md
Email Obfuscator AddOn für REDAXO 5
===================================

Durch dieses Addon werden alle Email-Adressen automatisch so verschleiert, dass sie von Spambots nicht mehr erkannt werden können. Dabei ist es egal ob die Email-Adressen sich in einem Template oder einem Block/Modul befinden.

Email-Adressen können mit oder ohne Anchor-Tag angegeben werden. Die folgenden Bespiele sind also möglich:

* `foo@gmx.de`
* `<a href="mailto:foo@gmx.de">Foo's EMail Adresse>`

Um die Email-Adressen zu schützen, werden die Techniken "CSS display:none" und "ROT13 Encryption" angewendet. Diese können weiter unten ein- oder ausgeschaltet werden. Weitere Informationen zu den Techniken in diesem Artikel: [Nine ways to obfuscate e-mail addresses compared](http://techblog.tilllate.com/2008/07/20/ten-methods-to-obfuscate-e-mail-addresses-compared)

__Wichtiger Hinweis:__ die CSS Display None Methode benötigt diesen Style: `span.hide { display: none; }`

Features
--------

* Vollautomatisches Verschleiern der Email-Adressen mit bewährten Algorithmen
* Sowohl nackte als auch Email-Adressen in einem A-Tag werden berücksichtigt
* Mehrere Verschleierungs-Methoden zur Auswahl

Hinweise
--------

* Getestet mit REDAXO 5.2
* Addon-Ordner lautet: `email_obfuscator`
* Die CSS Methode benötigt diesen Eintrag in Ihrem Stylesheet: `span.hide { display: none; }`

Changelog
---------

siehe CHANGELOG.md

Lizenz
------

MIT, siehe LICENSE.md

Credits
-------

* Danke an WordPress für die `make_clickable()` Funktion :)
* Danke an [Xong](https://github.com/xong) für die Hilfe zu RegEx


CHANGELOG.md
Email Obfuscator - Changelog
============================

### Version 2.0.0 - 15. März 2017

* Portierung zu REDAXO 5
* Templates und Artikel können vom Verschlüsseln ausgeschlossen werden
* Attribute in den a-Tags bleiben nun erhalten
* Bei leerer NoScript Message wird die Ausgabe von dem `noscript` Tag unterdrückt

### Version 1.6.0 - 25. Oktober 2014

* Updatefähigkeit für REDAXO 4.6 hergestellt. Einstellungen werden jetzt im Data-Ordner gespeichert.

### Version 1.5.1 - 23. August 2014

* Addon ab 1.5.1 nur noch für REDAXO 4.5+
* Späte OUTPUT_FILTER Registrierung, so dass andere Addons (opf_lang, String Table etc.) vorab Ersetzungen durchführen können, thx@ceekay82

### Version 1.5.0 - 25. Oktober 2013

* Fixed: Email-Adressen wurden innerhalb Input-Tags fälschlicherweise verschleiert
* Fixed: EMail-Adressen mit z.B. `.co.uk` Endung wurden nicht korrekt verschleiert
* Man kann in der `config.inc.php` den Hinweistext ändern, der erscheint wenn JavaScript deaktiviert ist.
* Man kann in der `config.inc.php` für die NoScript Meldung ein Key für das String Table Addon angeben (so mehrsprachiger Hinweistext möglich).

### Version 1.3.0 - 22. September 2013

* Fixed #7: Manche nackten Email-Adressen wurden nicht ersetzt insbesondere wenn diese zwischen einem HTML-Tag standen (z.B. `<p>foo@bar.de</p>`)

### Version 1.2.6 - 26. Mai 2013

* Ersetzung sollte jetzt auch bei einem Title-Attribut in Anchor-Tag sowie bei gleichlautendem Email- und Linktext funktionieren (#1)

### Version 1.2.5 - 13. März 2013

* AddOn Name geändert in `Email Obfuscator`
* License und Changelog hinzugefügt
* CSS Methode standardmäßig abgeschaltet

### Version 1.2.2 - 08. November 2012

* Hinweis für benötigten Style wenn man die CSS Methode nutzt ergänzt
* Unnötige Tags die für die XHTML Validierung nötig waren wurden entfernt

### Version 1.1.0 - 08. Juni 2012

* Der ausgespuckte Code validiert jetzt sauber
* Das Benutzerrecht protect_my_email[] wurde hinzugefügt

### Version 1.0.0 - 07. Dezember 2010




pages/config.php
<?php

$content = '';
$buttons = '';

// Einstellungen speichern
if (rex_post('formsubmit', 'string') == '1') {
    $this->setConfig(rex_post('config', [
        ['javascriptmethod', 'int'],
        ['nojavascriptmethod', 'int'],
        ['noscript_msg', 'string'],
		['articles', 'string'],
		['templates', 'array[int]']
    ]));

    echo rex_view::success($this->i18n('config_saved'));
}

// js metjod
$formElements = [];
$n = [];
$n['label'] = '<label for="javascriptmethod">' . $this->i18n('config_javascriptmethod') . '</label>';
$n['field'] = '<input type="checkbox" id="javascriptmethod" name="config[javascriptmethod]"' . ($this->getConfig('javascriptmethod') == '1' ? ' checked="checked"' : '') . ' value="1" />';
$formElements[] = $n;

$fragment = new rex_fragment();
$fragment->setVar('elements', $formElements, false);
$content .= $fragment->parse('core/form/checkbox.php');

// no js method
$formElements = [];
$n = [];
$n['label'] = '<label for="nojavascriptmethod">' . $this->i18n('config_nojavascriptmethod') . '</label>';
$n['field'] = '<input type="checkbox" id="nojavascriptmethod" name="config[nojavascriptmethod]"' . ($this->getConfig('nojavascriptmethod') == '1' ? ' checked="checked"' : '') . ' value="1" />';
$formElements[] = $n;

$fragment = new rex_fragment();
$fragment->setVar('elements', $formElements, false);
$content .= $fragment->parse('core/form/checkbox.php');

// js msg
$formElements = [];
$n = [];
$n['label'] = '<label for="noscript_msg">' . $this->i18n('config_noscript_msg') . '</label>';
$n['field'] = '<input class="form-control" type="text" id="noscript_msg" name="config[noscript_msg]" value="' . $this->getConfig('noscript_msg') . '"/>';
$formElements[] = $n;

$fragment = new rex_fragment();
$fragment->setVar('elements', $formElements, false);
$content .= $fragment->parse('core/form/container.php');


// whitelist stuff
$formElements = [];
$n = [];
$n['label'] = '<label for="email_obfuscator-config-articles">' . $this->i18n('config_articles') . '</label>';
$n['field'] = rex_var_linklist::getWidget(1, 'config[articles]', $this->getConfig('articles'));
$formElements[] = $n;

$n = [];
$n['label'] = '<label for="email_obfuscator-config-templates">' . $this->i18n('config_templates') . '</label>';
$select = new rex_select();
$select->setId('email_obfuscator-config-templates');
$select->setMultiple();
$select->setSize(10);
$select->setAttribute('class', 'form-control');
$select->setName('config[templates][]');
$select->addSqlOptions('SELECT `name`, `id` FROM `' . rex::getTablePrefix() . 'template` ORDER BY `name` ASC');
$select->setSelected($this->getConfig('templates'));
$n['field'] = $select->get();
$formElements[] = $n;

$fragment = new rex_fragment();
$fragment->setVar('elements', $formElements, false);
$content .= $fragment->parse('core/form/form.php');


// Save-Button
$formElements = [];
$n = [];
$n['field'] = '<button class="btn btn-save rex-form-aligned" type="submit" name="save" value="' . $this->i18n('config_save') . '">' . $this->i18n('config_save') . '</button>';
$formElements[] = $n;

$fragment = new rex_fragment();
$fragment->setVar('elements', $formElements, false);
$buttons = $fragment->parse('core/form/submit.php');
$buttons = '
<fieldset class="rex-form-action">
    ' . $buttons . '
</fieldset>
';

// Ausgabe Formular
$fragment = new rex_fragment();
$fragment->setVar('class', 'edit');
$fragment->setVar('title', $this->i18n('config'));
$fragment->setVar('body', $content, false);
$fragment->setVar('buttons', $buttons, false);
$output = $fragment->parse('core/page/section.php');

$output = '
<form action="' . rex_url::currentBackendPage() . '" method="post">
<input type="hidden" name="formsubmit" value="1" />
    ' . $output . '
</form>
';

echo $output;

?>



lib/rex_email_obfuscator.php
<?php
class rex_email_obfuscator {
	public static function obfuscate($ep) {
		$content = $ep->getSubject();
		$whitelistTemplates = rex_addon::get('email_obfuscator')->getConfig('templates', []);
		$whitelistArticles = rex_addon::get('email_obfuscator')->getConfig('articles', '');

		if ($whitelistArticles != '') {
			$whitelistArticles = explode(',', $whitelistArticles);
		} else {
			$whitelistArticles = [];
		}

		if (!is_null(rex_article::getCurrent()) && !in_array(rex_article::getCurrent()->getTemplateId(), $whitelistTemplates) && !in_array(rex_article::getCurrentId(), $whitelistArticles)) {
			$javascriptmethod = rex_addon::get('email_obfuscator')->getConfig('javascriptmethod');
			$nojavascriptmethod = rex_addon::get('email_obfuscator')->getConfig('nojavascriptmethod');
			$atPos = strpos($content, '@');
	
			if ($atPos === false || (!$javascriptmethod && !$nojavascriptmethod)) {
				// nothing to do
				return $content;
			}
	
			// wrap anchor tag around email-adresses that don't have already an anchor tag around them
			$content = self::makeEmailClickable($content);
	
			// replace all email addresses (now all wrapped in anchor tag) with spam aware version
			$content = preg_replace_callback('`\<a([^>]+)href\=\"mailto\:([^">]+)\"([^>]*)\>(.*?)\<\/a\>`ism', function ($m) {
				return self::encodeEmail($m[2], $m[4], $m[1], $m[3]);
			}, $content);
		
		}
	
		// done!
		return $content;
	}

	public static function encodeEmail($email, $text = "", $attributesBeforeHref = '', $attributesAfterHref = '') {
		if (empty($text)) {
			$text = $email;
		}

		$attributesBeforeHref = trim($attributesBeforeHref);
		$attributesAfterHref = trim($attributesAfterHref);

		if ($attributesBeforeHref != '') {
			$attributesBeforeHref = str_replace('"', '\\"', $attributesBeforeHref);
			$attributesBeforeHref = $attributesBeforeHref;
		}

		if ($attributesAfterHref != '') {
			$attributesAfterHref = str_replace('"', '\\"', $attributesAfterHref);
			$attributesAfterHref = ' ' . $attributesAfterHref;
		}
	
		$encoded = '';
		$javascriptmethod = rex_addon::get('email_obfuscator')->getConfig('javascriptmethod');
		$nojavascriptmethod = rex_addon::get('email_obfuscator')->getConfig('nojavascriptmethod');
	
		if ($javascriptmethod) {
			// javascript version
			$encoded_mail_tag = str_rot13('<a ' . $attributesBeforeHref . 'href=\\"mailto:' . $email . '\\"' . $attributesAfterHref . '>' . $text . '</a>');
			$encoded = "<script type=\"text/javascript\">";
			$encoded .= "/* <![CDATA[ */";
			$encoded .= "document.write(\"" . $encoded_mail_tag . "\".replace(/[a-zA-Z]/g, function(c){return String.fromCharCode((c<=\"Z\"?90:122)>=(c=c.charCodeAt(0)+13)?c:c-26);}));";
			$encoded .= "/* ]]> */";
			$encoded .= "</script>";
		}
	
		// for users who have javascript disabled
		$exploded_email = explode("@", $email);
	
		if ($javascriptmethod && !$nojavascriptmethod) {
			$noscriptMsg = rex_addon::get('email_obfuscator')->getConfig('noscript_msg');

			if ($noscriptMsg != '') {
				$encoded .= '<noscript><em>&gt;&gt;&gt; ' . $noscriptMsg . ' &lt;&lt;&lt;</em></noscript>';
			}
		} else {
			if ($javascriptmethod && $nojavascriptmethod) {
				$encoded .= '<noscript>';
			}

			// make cryptic strings
			$string_snippet = strtolower(str_rot13(preg_replace("/[^a-zA-Z]/", "", $email)));
			$cryptValues = str_split($string_snippet, 5);
		
			if ($nojavascriptmethod) {
				$encoded .= "<span class=\"hide\">" . $cryptValues[0] . "</span>" . $exploded_email[0] . "<span class=\"hide\">" . strrev($cryptValues[0]) . "</span>[at]<span class=\"hide\">" . $cryptValues[0] . "</span>" . $exploded_email[1];
			}
		
			if ($javascriptmethod && $nojavascriptmethod) {
				$encoded .= '</noscript>';
			}
		}
	
		return $encoded;
	}

	// found here: http://zenverse.net/php-function-to-auto-convert-url-into-hyperlink/
	public static function makeEmailClickable($ret) {
		$ret = ' ' . $ret;
		// in testing, using arrays here was found to be faster
		$ret = preg_replace_callback('#([\s>])([.0-9a-z_+-]+)@(([0-9a-z-]+\.)+[0-9a-z]{2,})#i', 'rex_email_obfuscator::_make_email_clickable_cb', $ret);
	 
		// this one is not in an array because we need it to run last, for cleanup of accidental links within links
		$ret = preg_replace("#(<a( [^>]+?>|>))<a [^>]+?>([^>]+?)</a></a>#i", "$1$3</a>", $ret);
		$ret = trim($ret);

		return $ret;
	}

	protected static function _make_email_clickable_cb($matches) {
		$email = $matches[2] . '@' . $matches[3];

		return $matches[1] . "<a href=\"mailto:$email\">$email</a>";
	}
}
